### ISSUE TITLE
Tests not fonctionnal

### ISSUE THREAD

**Issue by fredjarlier** [marcnol/pyHiM#318]
Hi,
I would like to test the tool but I have a problem  with pytest. After cloning I do:
 
git submodule update --init --recursive

but the following link doesn't exist
 
https://github.com/pyHi-M/pyHiM/blob/development/pyhim-small-dataset

is it normal?

Best

fred

**Comment by marcnol** [comment]
Hi Fred, no this is not normal. We will get back to you soon with a fix.

**Comment by XDevos** [comment]
You can solve it by doing : 
`git clone --recurse-submodules https://github.com/pyHi-M/pyhim-small-dataset.git`

And then run:
```bash
cd ~Repositories/pyHiM/
conda activate pyhim39
pytest tests/ -vv
```

**Comment by fredjarlier** [comment]
Hi @XDevos and @marcnol , 

Indeed the clone recursive works thanks.

We happen to create the following environment in a singularity container:

 git clone https://github.com/pyHi-M/traceratops.git
 cd /opt/Repositories/traceratops/
 git checkout dev
 conda run -n pyhim39 pip install -e ".[dev]"
 cd /opt/Repositories
 git clone https://github.com/apiFISH/apiFISH.git
 cd apiFISH && git checkout development
 conda run -n pyhim39 pip install -e ".[dev]"
 cd /opt/Repositories/pyHiM
 conda run -n pyhim39 pip install -e ".[dev]"
 git clone --recurse-submodules https://github.com/pyHi-M/pyhim-small-dataset.git

but when we run in the container

pytest tests/ -vv 

we finish with 

========== 13 failed, 10 passed, 46 warnings in 1674.36s (0:27:54) =====

do you think it's normal? should'nt all tests success?

best

fred

**Comment by fredjarlier** [comment]
Hi @marcnol 

Here is the toal environment i'm testing

    cd /opt/Repositories
    git clone https://github.com/pyHi-M/pyHiM.git
    cd pyHiM
    git checkout development
    cp /opt/nodefault_environment.yml /opt/Repositories/pyHiM/
    conda env create -f environment.yml
    cd /opt/Repositories
    git clone https://github.com/pyHi-M/traceratops.git
    cd /opt/Repositories/traceratops/
    git checkout dev
    conda run -n pyhim39 pip install -e ".[dev]"
    cd /opt/Repositories
    git clone https://github.com/apiFISH/apiFISH.git
    cd apiFISH && git checkout development
    conda run -n pyhim39 pip install -e ".[dev]"
    cd /opt/Repositories/pyHiM
    conda run -n pyhim39 pip install -e ".[dev]"

and the environment I'm using for pyHim

name: pyhim39
channels:
  - nodefaults
  - bioconda
  - conda-forge
  - services-build-deps
dependencies:
  - _libgcc_mutex=0.1
  - _openmp_mutex=5.1
  - ca-certificates
  - ld_impl_linux-64
  - libffi
  - libgcc-ng
  - libgomp
  - libstdcxx-ng
  - ncurses
  - openssl
  - pip
  - python=3.9.21
  - readline
  - setuptools
  - sqlite
  - tk
  - wheel
  - xz
  - zlib
  - pip:
      - absl-py==2.2.0
      - astropy==6.0.1
      - astropy-iers-data==0.2025.3.17.0.34.53
      - astunparse==1.6.3
      - certifi==2025.1.31
      - charset-normalizer==3.4.1
      - click==8.1.8
      - cloudpickle==3.1.1
      - csbdeep==0.8.1
      - cycler==0.12.1
      - dask==2024.8.0
      - dataclasses-json==0.6.7
      - distributed==2024.8.0
      - exceptiongroup==1.2.2
      - flatbuffers==25.2.10
      - fonttools==4.56.0
      - fsspec==2025.3.0
      - gast==0.6.0
      - google-pasta==0.2.0
      - grpcio==1.71.0
      - h5py==3.13.0
      - idna==3.10
      - imageio==2.37.0
      - importlib-metadata==8.6.1
      - iniconfig==2.1.0
      - jinja2==3.1.6
      - joblib==1.4.2
      - keras==3.9.0
      - kiwisolver==1.4.7
      - libclang==18.1.1
      - llvmlite==0.43.0
      - locket==1.0.0
      - markdown==3.7
      - markdown-it-py==3.0.0
      - markupsafe==3.0.2
      - marshmallow==3.26.1
      - matplotlib==3.5.1
      - mdurl==0.1.2
      - ml-dtypes==0.5.1
      - mrc==0.3.1
      - msgpack==1.1.0
      - mypy-extensions==1.0.0
      - namex==0.0.8
      - networkx==3.2.1
      - numba==0.60.0
      - numpy==1.26.4
      - opt-einsum==3.4.0
      - optree==0.14.1
      - packaging==24.2
      - pandas==2.2.3
      - partd==1.4.2
      - photutils==1.11.0
      - pillow==11.1.0
      - pluggy==1.5.0
      - protobuf==5.29.4
      - psutil==7.0.0
      - pyerfa==2.0.1.5
      - pygments==2.19.1
      - pympler==1.1
      - pyparsing==3.2.1
      - pytest==8.3.5
      - python-dateutil==2.9.0.post0
      - pytz==2025.1
      - pywavelets==1.6.0
      - pyyaml==6.0.2
      - requests==2.32.3
      - rich==13.9.4
      - roipoly==0.5.3
      - scikit-image==0.19.2
      - scikit-learn==1.6.1
      - scipy==1.13.1
      - six==1.17.0
      - sortedcontainers==2.4.0
      - stardist==0.9.1
      - tblib==3.0.0
      - tensorboard==2.19.0
      - tensorboard-data-server==0.7.2
      - tensorflow==2.19.0
      - tensorflow-addons==0.23.0
      - tensorflow-io-gcs-filesystem==0.37.1
      - termcolor==2.5.0
      - threadpoolctl==3.6.0
      - tifffile==2024.8.30
      - tomli==2.2.1
      - toolz==1.0.0
      - tornado==6.4.2
      - tqdm==4.67.1
      - typeguard==2.13.3
      - typing-extensions==4.12.2
      - typing-inspect==0.9.0
      - tzdata==2025.1
      - urllib3==2.3.0
      - werkzeug==3.1.3
      - wrapt==1.17.2
      - zict==3.0.0
      - zipp==3.21.0

as you can see I trade default to  nodefault

Could you test with this environment and tell me please?

best

fred

**Comment by marcnol** [comment]
Hi Fred, installed pyHiM into a new computer that has never seen it. I followed your same steps, and also the tests do not pass. I used the `default` channel to avoid issues coming from change in channels.

Here is the message I get. I will dig into this with Xavier to know why the tests are not passing.

In any case, you can from now move on and test it with the test dataset I shared.
 

```bash
(pyhim39) marcnol@sysadm-XPS-13-9360:~/Repositories/pyHiM$ pytest tests/ -vv

=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.9.21, pytest-8.3.5, pluggy-1.5.0 -- /home/marcnol/miniforge3/envs/pyhim39/bin/python
cachedir: .pytest_cache
rootdir: /home/marcnol/Repositories/pyHiM
configfile: pyproject.toml
plugins: typeguard-2.13.3
collected 23 items                                                                                                                                                                                                

tests/test_core/test_run_args.py::test_check_consistency PASSED                                                                                                                                             [  4%]
tests/test_localize_2d.py::test_inhomogeneous FAILED                                                                                                                                                        [  8%]
tests/test_localize_3d.py::test_full_stack_stardist Fatal Python error: Aborted

Thread 0x00007d7587a2c440 (most recent call first):
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/execute.py", line 53 in quick_execute
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/context.py", line 1688 in call_function
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/polymorphic_function/atomic_function.py", line 251 in call_flat
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/polymorphic_function/atomic_function.py", line 216 in call_preflattened
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/polymorphic_function/concrete_function.py", line 1322 in _call_flat
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/polymorphic_function/polymorphic_function.py", line 919 in _call
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/eager/polymorphic_function/polymorphic_function.py", line 833 in __call__
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/tensorflow/python/util/traceback_utils.py", line 150 in error_handler
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/keras/src/backend/tensorflow/trainer.py", line 560 in predict
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 117 in error_handler
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/stardist/models/base.py", line 409 in predict_direct
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/stardist/models/base.py", line 603 in _predict_sparse_generator
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/stardist/models/base.py", line 740 in _predict_instances_generator
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/stardist/models/base.py", line 788 in predict_instances
  File "/home/marcnol/Repositories/pyHiM/src/imageProcessing/segmentMasks.py", line 950 in _segment_3d_volumes_stardist
  File "/home/marcnol/Repositories/pyHiM/src/imageProcessing/segmentSources3D.py", line 175 in _segment_3d_volumes
  File "/home/marcnol/Repositories/pyHiM/src/imageProcessing/segmentSources3D.py", line 274 in segment_sources_3d_file
  File "/home/marcnol/Repositories/pyHiM/src/imageProcessing/segmentSources3D.py", line 469 in segment_sources_3d_in_folder
  File "/home/marcnol/Repositories/pyHiM/src/imageProcessing/segmentSources3D.py", line 536 in segment_sources_3d
  File "/home/marcnol/Repositories/pyHiM/src/core/function_caller.py", line 432 in segment_sources_3d
  File "/home/marcnol/Repositories/pyHiM/src/pyHiM.py", line 134 in main
  File "/home/marcnol/Repositories/pyHiM/tests/test_localize_3d.py", line 32 in template_test_localize_3d
  File "/home/marcnol/Repositories/pyHiM/tests/test_localize_3d.py", line 66 in test_full_stack_stardist
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/python.py", line 159 in pytest_pyfunc_call
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/python.py", line 1627 in runtest
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/runner.py", line 174 in pytest_runtest_call
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/runner.py", line 242 in <lambda>
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/runner.py", line 341 in from_call
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/runner.py", line 241 in call_and_report
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/runner.py", line 132 in runtestprotocol
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/runner.py", line 113 in pytest_runtest_protocol
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/main.py", line 362 in pytest_runtestloop
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/main.py", line 337 in _main
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/config/__init__.py", line 175 in main
  File "/home/marcnol/miniforge3/envs/pyhim39/lib/python3.9/site-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/home/marcnol/miniforge3/envs/pyhim39/bin/pytest", line 8 in <module>
Aborted (core dumped)

```

Here is my terminal history for the install:

```bash
   59  cd Repositories/
   66  git clone https://github.com/pyHi-M/pyHiM.git
   67  cd pyHiM/
   69  git checkout development
   71  nano $HOME/.bashrc
   72  mamba env create -f environment.yml
   73  cd ..
   61  git clone https://github.com/pyHi-M/traceratops.git
   62  cd traceratops/
   63  git checkout dev
   75  mamba run -n pyhim39 pip install -e ".[dev]"
   76  cd ..
   77  git clone https://github.com/apiFISH/apiFISH.git
   78  cd apiFISH && git checkout development
   79  mamba run -n pyhim39 pip install -e ".[dev]"
   80  cd ../pyHiM/
   82  git clone --recurse-submodules https://github.com/pyHi-M/pyhim-small-dataset.git
   83  mamba activate pyhim39
   84  pytest tests/ -vv
```

**Comment by fredjarlier** [comment]
Hi @marcnol, 

Yes I'm starting the tests with data you share: rawData_2022_Exp6_ROI5.
But I have another trouble with Dask and singularity. I'm looking a way around...I keep you in touch.

best

fred

**Comment by marcnol** [comment]
I have worked this week on making a version of pyHiM that can launch one file at a time, and that does therefore not use dask. I will do a PR later on the week once all tests are safe
